spec.md

```diff
@@ -3690,6 +3738,9 @@ statement =
   | "connect" , reference , "," , expr , [ info ]
   | "invalidate" , reference , [ info ]
 
+(* Group definitions *)
+defgroup = "defgroup" , id ;
```

Comment:
What is this top level one for? Can you give an example?

Reply:
It's to declare the groups that can be used later. This makes it way easier to know what groups exist. Otherwise, you need to scan the entire IR and look for any group instances. This is also then used for check if a group instance is legal. If circuit `Foo` is conforming to an interface, it can only have certain legal groups. By making the groups declared up front this is easier.

Here's a bigger example. It would be illegal for any of the modules to have a "wildcat" group like `group qux of Qux` because that was not declared.

```
circuit Foo:
  defgroup Bar           ; declare the group Bar

  module Baz:
    input a: UInt<1>

    group bar of Bar:    ; create the group instance bar
      node notA = not(a)

  module Foo:
    input a: UInt<1>

    group bar of Bar:    ; create the group instance bar
      node notA = not(a)

    inst baz of Baz
    baz.a <= a
```

This will then produce the following Verilog:

```verilog
// Inside file "Foo.sv":
module Foo(
  input a
  Baz baz(.a(a));
);
endmodule


// Inside file "Baz.sv":
module Baz(
  input a
);
endmodule

// Inside file "Foo_Bar.sv":
module Foo_Bar(
  input a
);
  wire notA = ~a;
endmodule

// Inside file "Baz_Bar.sv":
module Baz_Bar(
  input a
);
  wire notA = ~a;
endmodule

// Inside file "bindings_Foo_Bar.sv" :
bind Foo Foo_Bar bar(.a(a));
bind Baz Baz_Bar bar(.a(a));
```